rfc/rfc004-verifiable-transactional-graph.md

```diff
@@ -107,25 +100,18 @@ The following orders are invalid:
 * `A -> B -> D -> F -> C -> E` \(merger processed before all previous were processed\)
 * `A -> B -> D -> E -> C -> F` \(branch C is processed out-of-order\)
 
-### 3.4. Timelines
-
-Since transactions are immutable, the only way to update the application data they contain is by creating a new transaction. Subsequent versions of the transaction's application data SHOULD be tracked by creating a _timeline_ using the optional **tid** and **tiv** fields. These fields SHALL NOT be used on the first transaction in the timeline, only on updates. The **tid** field identifies the timeline and MUST contain the reference to the first transaction. The **tid** field MUST be present when **tiv** is specified.
-
-For signalling updates based on out-of-date state **tiv** \(timeline version\) incrementing integer CAN be used to indicate the version of the application data. The first transaction in the timeline MUST have a **tiv** of `0`, but since this is the default value it CAN be omitted. The first update in the timeline MUST have a **tiv** of `1`, the second `2` and so on.
-
-Incorrect state due to incorrect updates \(e.g. duplicate **tiv**, see below\) SHOULD be fixed by issuing a new update with incremented **tiv**.
-
-#### 3.4.1. Timeline validation
+### 3.4. Updates & Ordering
 
-It's up to the processing application to validate the timeline, but the following points SHOULD be taken into consideration:
+Since transactions are immutable, the only way to update the application data they contain is by creating a new transaction. 
+All updates are considered as full updates, partial updates are not supported. 
+Parallel updates of the same application data, in the form of branches, MUST be processed the same way by all nodes.
 
-* Assert that the update has been issued by the owner \(signer\) of the original transaction.
-* Gaps in timeline versioning MAY indicate the consumer is missing a \(branching\) transaction.
-* Duplicates in timeline versioning MAY indicate a race condition in which the producer updated the application data based on out-of-date state.
-  * When the payloads are equal one of the updates COULD be applied and the other one ignored.
-  * When the payloads differ the transaction with the lowest **iat** SHOULD be applied first.
+Consider the DAG from the previous chapter. If transaction `C` and  `D` where to update the same application data, nodes could process the transaction in a different order.
+This would create an inconsistency in the network. To fix this the following rules MUST be taken into account:
 
-    If **iat**s are equal the transaction with the lowest hash should be applied first.
+* If the payloads are equal, mark both transactions as processed and apply one of the payloads.
```

Comment:
This change means we have to (severely) revise the DAG walking algorithm.

Also a complication: since the RFC004 layer now doesn't know what transactions are about the same "application data" (previously it did through `timeline ID`), the algorithm can't determine what transactions should be taken into consideration to determine the order.

Another problem which exists in the current implementation as well but doesn't get fixed with this change is that the transaction ordering (of already applied transactions)  might change:
When a new transaction with a older `iat` gets published as branch of an older transaction, the transactions might have to be replayed. This could be a problem when DID `authentication` keys are removed, which might invalidate future versions of that DID document which were previously considered valid.

Not sure what the solution is without making DAG traversal when history changes massively complicated.

We might push that responsibility into the application layer: VDR could "insert" that version where it belongs (according to `iat`) and validate all succeeding versions from that point on? Complication is that information derived from DIDs (VCs) could be impacted/invalidated too, and should be reprocessed as well, further complicating things.

On the other hand, since every update is a full replacement it might not matter that much.

Reply:
I see the point. VCs might have been issued with a key that is no longer valid a few seconds later when a new DID Document update has been processed. But don't you have the same problem when two transactions use the same timeline version number. According to the current rules a similar sorting applies?

Reply:
And it's only a problem for docs with the exact same `iat`

Reply:
@reinkrul I changed the line. So we don't have to change the DAG algorithm (yet). All other remarks are unrelated to this PR?